# Data Extraction and EDA

This development notebook outlines the steps for data extraction and exploratory data analysis (EDA). The dataset is stored in a local folder named "data/raw," containing two CSV files representing products from Retailer A and Retailer B. Our goal is to extract and explore these datasets in preparation for entity matching and clustering.

This notebook contains code snippets that will be included in the `data_extraction.py` and `data_processing.py` scripts. These scripts are integral components of an automated pipeline designed for data extraction and processing.

In [11]:
# Imports for data extraction and EDA
import pandas as pd

## Reading RetalierA and RetailerB

In this section we will read the Retailer A and Retailer B data separately and later merge them to have a combined dataset that can be useful for later analysis and processing.

In [12]:
# Retailer A data
retA_data = pd.read_csv('../data/raw/retailerA.csv')
retA_data.head()

unique_id                                              title  \
0          1   Linksys EtherFast 8-Port 10/100 Switch - EZXS88W   
1          2  Linksys EtherFast10/100 5-Port Auto-Sensing Sw...   
2          3  Netgear ProSafe 5 Port 10/100 Desktop Switch -...   
3          4  Belkin F3H982-10 Pro Series High Integrity 10 ...   
4          5  Netgear Prosafe 16 Port 10/100 Rackmount Switc...   

                                         description    price  
0  Linksys EtherFast 8-Port 10/100 Switch - EZXS8...   $44.00  
1  Linksys EtherFast10/100 5-Port Auto-Sensing Sw...   $29.00  
2  Netgear ProSafe 5 Port 10/100 Desktop Switch -...   $40.00  
3  Belkin F3H982-10 Pro Series High Integrity 10 ...      NaN  
4  Netgear Prosafe 16 Port 10/100 Rackmount Switc...  $131.00

In [13]:
retA_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1081 entries, 0 to 1080
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   unique_id    1081 non-null   int64 
 1   title        1081 non-null   object
 2   description  1081 non-null   object
 3   price        418 non-null    object
dtypes: int64(1), object(3)
memory usage: 33.9+ KB


In [14]:
# Retailer A data
retB_data = pd.read_csv('../data/raw/retailerB.csv')
retB_data.head()

unique_id                                              title  \
0          1  Linksys EtherFast EZXS88W Ethernet Switch - EZ...   
1          2          Linksys EtherFast EZXS55W Ethernet Switch   
2          3    Netgear ProSafe FS105 Ethernet Switch - FS105NA   
3          4  Belkin Pro Series High Integrity VGA/SVGA Moni...   
4          5             Netgear ProSafe JFS516 Ethernet Switch   

                                         description price  
0  Linksys EtherFast 8-Port 10/100 Switch (New/Wo...   NaN  
1                              5 x 10/100Base-TX LAN   NaN  
2  NETGEAR FS105 Prosafe 5 Port 10/100 Desktop Sw...   NaN  
3               1 x HD-15 - 1 x HD-15 - 10ft - Beige   NaN  
4  Netgear ProSafe 16 Port 10/100 Rackmount Switc...   NaN

In [15]:
retB_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1092 entries, 0 to 1091
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   unique_id    1092 non-null   int64 
 1   title        1092 non-null   object
 2   description  646 non-null    object
 3   price        590 non-null    object
dtypes: int64(1), object(3)
memory usage: 34.3+ KB


## Datasets (A and B) Structure

| Column Name  | Data Type      | Description                                       |
|--------------|----------------|---------------------------------------------------|
| `unique_id`    | Numeric (int)  | Unique identifier for each product record.       |
| `title`        | Text (string)  | Title or name of the product.                    |
| `description`  | Text (string)  | Description of the product.                      |
| `price`     | Numeric (float)| Price of the product.                            |

This table provides an overview of each column in the dataset along with its respective data type and a brief description of its content.

In [18]:
# Combining data from RetA and RetB
combined_data = pd.concat([retA_data, retB_data])
combined_data.head()

unique_id                                              title  \
0          1   Linksys EtherFast 8-Port 10/100 Switch - EZXS88W   
1          2  Linksys EtherFast10/100 5-Port Auto-Sensing Sw...   
2          3  Netgear ProSafe 5 Port 10/100 Desktop Switch -...   
3          4  Belkin F3H982-10 Pro Series High Integrity 10 ...   
4          5  Netgear Prosafe 16 Port 10/100 Rackmount Switc...   

                                         description    price  
0  Linksys EtherFast 8-Port 10/100 Switch - EZXS8...   $44.00  
1  Linksys EtherFast10/100 5-Port Auto-Sensing Sw...   $29.00  
2  Netgear ProSafe 5 Port 10/100 Desktop Switch -...   $40.00  
3  Belkin F3H982-10 Pro Series High Integrity 10 ...      NaN  
4  Netgear Prosafe 16 Port 10/100 Rackmount Switc...  $131.00

In [19]:
combined_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2173 entries, 0 to 1091
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   unique_id    2173 non-null   int64 
 1   title        2173 non-null   object
 2   description  1727 non-null   object
 3   price        1008 non-null   object
dtypes: int64(1), object(3)
memory usage: 84.9+ KB


## Combined Dataset Structure

| Column Name  | Data Type         | Description                                             | Number of Null Values | Number of Total Records | Remarks |
|--------------|-------------------|---------------------------------------------------------|-----------------------|-------------------------|-----------------------|
| `unique_id`   | Numeric (int64)   | Unique identifier for each product record.              | 0                     | 2173                    | None                  |
| `title`        | Text (object)     | Title or name of the product.                           | 0                     | 2173                    | None                  |
| `description`  | Text (object)     | Description of the product.                             | 446                   | 2173                    | Some descriptions are missing. |
| `price`        | Text (object)     | Price of the product.                                   | 1165                  | 2173                    | Many price values are missing. |

This table provides a comprehensive overview of the dataset, highlighting the columns, their data types, descriptions, the number of null values in each column, the total number of records, and relevant information about the dataset. It is evident that the `description` and `price` columns have a significant number of missing values.

## Feature Selection

Based on our preliminary data assessment, and for the scope of this case study, we have decided to focus exclusively on the `title` column of the dataset. This column provides a comprehensive description of the product and does not contain any missing values. Therefore, for the development of this case study, we will exclusively utilize the `title` column as it adequately serves our purposes.

In [ ]:
# Intermediate save for development purpose
combined_data.to_csv('../data/raw/combined_data_raw.csv')

----

### **AI tool usage for this notebook**

#### ChatGPT 3.5
* Improving makrdown annotations and table descriptions